modules/generator/registry/counter.go

```diff
@@ -0,0 +1,140 @@
+package registry
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"go.uber.org/atomic"
+)
+
+type counter struct {
+	name   string
+	labels []string
+
+	// seriesMtx is used to sync modifications to the map, not to the data in series
+	seriesMtx sync.RWMutex
+	series    map[uint64]*counterSeries
+
+	onAddSeries    func(count uint32) bool
+	onRemoveSeries func(count uint32)
+}
+
+type counterSeries struct {
+	// labelValues should not be modified after creation
+	labelValues []string
+	value       *atomic.Float64
+	lastUpdated *atomic.Int64
+}
+
+var _ Counter = (*counter)(nil)
+var _ metric = (*counter)(nil)
+
+func newCounter(name string, labels []string) *counter {
+	return &counter{
+		name:   name,
+		labels: labels,
+		series: make(map[uint64]*counterSeries),
+	}
+}
+
+func (c *counter) setCallbacks(onAddSeries func(uint32) bool, onRemoveSeries func(count uint32)) {
+	c.onAddSeries = onAddSeries
+	c.onRemoveSeries = onRemoveSeries
+}
+
+func (c *counter) Inc(labelValues *LabelValues, value float64) {
+	if value < 0 {
+		panic("counter can only increase")
```

Comment:
Should we panic here?  Would an error log and early return be sufficient?

Reply:
I think it's ok, the Prometheus client [does the same](https://github.com/prometheus/client_golang/blob/main/prometheus/counter.go#L107), although the method is more general `Add` vs `Inc` which takes no params.

Reply:
Yeah, we could handle this a bit more elegantly, but I didn't want to bother adding a `log.Logger` field to the counter or returning an error just in case someone misuses the counter.
A panic isn't nice, but it should ensure whoever tries to decrease the counter detects it before it ships.

Prometheus client_golang also just panics btw: https://github.com/prometheus/client_golang/blob/main/prometheus/counter.go#L108-L110